### PR TITLE
CURATOR-623 - ChildrenCache (used by Queues) didn't have a ConnectionStateListener

### PR DESCRIPTION
**Pull request by Randgalt**
`ChildrenCache` (used by Queues) didn't have a `ConnectionStateListener`. Thus, if a long network partition occurred the ZK instance would be recreated losing any set watcher and the ChildrenCache would fail to continue watching changes. Adding a ConnectionStateListener fixes this.

### CODE REVIEW

File: curator-recipes/src/main/java/org/apache/curator/framework/recipes/queue/ChildrenCache.java [hunk 1]
```diff
@@ -86,12 +101,14 @@ private Data(List<String> children, long version)
...
     {
-        sync(true);
+        client.getConnectionStateListenable().addListener(connectionStateListener);
+        sync();
     }
 
     @Override
     public void close() throws IOException
     {
+        client.getConnectionStateListenable().removeListener(connectionStateListener);
```
Comment (faucct) [inline curator-recipes/src/main/java/org/apache/curator/framework/recipes/queue/ChildrenCache.java:111]: Should not this be after `client.removeWatchers()`, in reverse order with `#start()`?

File: curator-recipes/src/main/java/org/apache/curator/framework/recipes/queue/ChildrenCache.java [hunk 2]
```diff
@@ -66,6 +68,19 @@ public void processResult(CuratorFramework client, CuratorEvent event) throws Ex
...
 
+    private final ConnectionStateListener connectionStateListener = (__, newState) -> {
+        if ((newState == ConnectionState.CONNECTED) || (newState == ConnectionState.RECONNECTED)) {
+            try
+            {
+                sync();
+            }
+            catch ( Exception e )
+            {
+                throw new RuntimeException(e);
```
Comment (eolivelli) [inline curator-recipes/src/main/java/org/apache/curator/framework/recipes/queue/ChildrenCache.java:79]: will this error be logged somewhere ?
as we cannot perform any recovery
will it make more sense to simply log the exception and do not rethrow it ?

**Review by eolivelli — CHANGES_REQUESTED** [review]
overall LGTM

I left one comment

File: curator-recipes/src/main/java/org/apache/curator/framework/recipes/queue/ChildrenCache.java (hunk 2 above)
Comment (Randgalt) [reply curator-recipes/src/main/java/org/apache/curator/framework/recipes/queue/ChildrenCache.java:79]: Yeah, it logs. It's not very obvious but these are handled by `MappingListenerManager` (see https://github.com/apache/curator/blob/master/curator-framework/src/main/java/org/apache/curator/framework/listen/MappingListenerManager.java#L85)

**Review by cammckenzie — APPROVED** [review]

**Review by eolivelli — APPROVED** [review]
LGTM

**Comment by eolivelli** [comment]
CI failed, please take a look:

[ERROR] org.apache.curator.framework.state.TestConnectionStateManager.testConnectionStateRecoversFromUnexpectedExpiredConnection
[ERROR]   Run 1: TestConnectionStateManager.testConnectionStateRecoversFromUnexpectedExpiredConnection:110 NoSuchMethod
[ERROR]   Run 2: TestConnectionStateManager.testConnectionStateRecoversFromUnexpectedExpiredConnection:110 NoSuchMethod
[ERROR]   Run 3: TestConnectionStateManager.testConnectionStateRecoversFromUnexpectedExpiredConnection:110 NoSuchMethod

**Comment by Randgalt** [comment]
> CI failed, please take a look:
> 
> [ERROR] org.apache.curator.framework.state.TestConnectionStateManager.testConnectionStateRecoversFromUnexpectedExpiredConnection [ERROR] Run 1: TestConnectionStateManager.testConnectionStateRecoversFromUnexpectedExpiredConnection:110 NoSuchMethod [ERROR] Run 2: TestConnectionStateManager.testConnectionStateRecoversFromUnexpectedExpiredConnection:110 NoSuchMethod [ERROR] Run 3: TestConnectionStateManager.testConnectionStateRecoversFromUnexpectedExpiredConnection:110 NoSuchMethod

The tests pass locally when I run them. Curator has so much flakiness with testing. I'll re-run Travis to see what it shows.

**Comment by shikharid** [comment]
Any plans to merge and release this?
And will this fix be backported to 2.xx release?
I recently faced this where a network blip of couple of minutes left all consumers of the queue stopped receiving any updates. In case this won't be backported, I will have to create a fork and apply those changes for the time being.

**Comment by tisonkun** [comment]
@Randgalt I've merge master into this branch and tests passed. Would you like to proceed this patch?

**Comment by Randgalt** [comment]
Wow - I don't have any memory of writing this :D I'll try to look at it when I get some time otherwise I trust your judgement on it.

**Review by tisonkun — APPROVED** [review]

**Comment by tisonkun** [comment]
@Randgalt LGTM. I'll merge it at this weekend if there is no further concern.

**Pull request merged**